5.4: Add documentation for the session changes (#107)
WoltLab/woltlab.github.io: pages/migration/wsc-53/migration_wsc-53_session.md

The legacy persistent login was rather an automated login.
Upon bootstrapping a session it was checked whether the user had a cookie pair storing the user’s userID and (a single BCrypt hash of) the user’s password.
If such a cookie pair exists and the single BCrypt hash hashes to the stored password hash then the session would immediately `changeUser()` to the respective user.

This legacy persistent login was completely removed.
Instead any sessions that belong to an authenticated user will automatically be long-lived.
These long-lived sessions expire no sooner than 14 days after the last activity, ensuring that the user continously stays logged in, provided that they visit the page at least once per fortnight.

### Multiple Sessions

To allow for a proper separation of these long-lived user sessions WoltLab Suite now allows for multiple sessions per user.
These sessions are completely unrelated to each other.
Specifically they do not share session variables and they expire independently.

As the existing `wcf1_session` table is also used for the online lists and location tracking it will be maintained on a best effort basis.
It no longer stores any private session data.

> The actual sessions storing security sensitive information are stored in an unrelated location.

Author: I intentionally opted for location, because the tables are off limits anyway and I plan to add Redis support in the future.

They must only be accessed via the PHP API exposed by the `SessionHandler`.

### Improved authentication and reauthentication

WoltLab Suite 5.4 ships with multi-factor authentication support and a generic re-authentication implementation that can be used to verify the account owner’s presence.

## Additions and Changes

### Password Hashing

WoltLab Suite 5.4 includes a new object oriented password hashing framework that is modeled after PHP’s `password_*` API.
Check [`PasswordAlgorithmManager`](https://github.com/WoltLab/WCF/blob/master/wcfsetup/install/files/lib/system/user/authentication/password/PasswordAlgorithmManager.class.php) and [`IPasswordAlgorithm`](https://github.com/WoltLab/WCF/blob/master/wcfsetup/install/files/lib/system/user/authentication/password/IPasswordAlgorithm.class.php) for details.

The new default password hash is a standard BCrypt hash.

> All newly generated hashes in `wcf1_user.password` will include a type prefix, instead of special casing the default hash.

Author: Is it actually? The fun part about English is that you can verbify all the nouns and the result can be easily understood. In fact “Special casing” as a verb is even [mentioned in Wiktionary](https://en.wiktionary.org/wiki/special_case#Verb) and I'd say it's somewhat common in software development jargon.

> All newly generated hashes in `wcf1_user.password` will include a type prefix, instead of special casing the default hash.

Reviewer: I still think that it is too obscure. Unless there is a very specific reason to use uncommon word, we should fall back to easier words to decrease the difficulty of the text for non-native English speakers.

> All newly generated hashes in `wcf1_user.password` will include a type prefix, instead of special casing the default hash.

Author: ```suggestion
All newly generated hashes in `wcf1_user.password` will now include a type prefix, instead of just passwords imported from other systems.
```

```

After:

```php
$isPending = WCF::getSession()->changeUserAfterMultifactorAuthentication($targetUser);
if ($isPending) {
	// Redirect to the authentication form. The user will not be logged in.
	// Note: Do not use `getControllerLink` to support both the frontend as well as the ACP.
	HeaderUtil::redirect(LinkHandler::getInstance()->getLink('MultifactorAuthentication', [
		'url' => /* Return To */,
	]);
	exit;
}
// Proceed as usual. The user will be logged in.
```

#### Adding Multi-factor Methods

Adding your own multi-factor method requires the implementation of a single object type:

```xml
<type>
	<name>com.example.multifactor.foobar</name>
	<definitionname>com.woltlab.wcf.multifactor</definitionname>
	<icon><!-- Font Awesome 4 Icon Name goes here. --></icon>
	<priority><!-- Determines the sort order, higher priority will be preferred for authentication. --></priority>
	<classname>wcf\system\user\multifactor\FoobarMultifactorMethod</classname>
</type>
```

The given classname must implement the [`IMultifactorMethod`](https://github.com/WoltLab/WCF/blob/master/wcfsetup/install/files/lib/system/user/multifactor/IMultifactorMethod.class.php) interface.

> As a self-contained example you can find the initial implementation of the email multi-factor method in [WoltLab/WCF#3729](https://github.com/WoltLab/WCF/pull/3729).

Reviewer: ```suggestion
As a self-contained example, you can find the initial implementation of the email multi-factor method in [WoltLab/WCF#3729](https://github.com/WoltLab/WCF/pull/3729).
```

Please check [the version history](https://github.com/WoltLab/WCF/commits/master/wcfsetup/install/files/lib/system/user/multifactor/EmailMultifactorMethod.class.php) of the PHP class to make sure you do not miss important changes that were added later.

{% include callout.html content="Multi-factor authentication is security sensitive.
Make sure to carefully read the remarks in IMultifactorMethod for possible issues.
Also make sure to carefully test your implementation against all sorts of incorrect input and consider attack vectors such as race conditions.
It is strongly recommended to generously check the current state by leveraging assertions and exceptions." type="warning" %}

## Deprecations and Removals

### SessionHandler

Most of the changes with regard to the new session handling happened in `SessionHandler`.

> Most notably `SessionHandler` now is marked `final` to ensure proper encapsulation of data.

Reviewer: ```suggestion
Most notably, `SessionHandler` now is marked `final` to ensure proper encapsulation of data.
```

> It is not expected that anyone inherited from `SessionHandler`, as it can be considered an implementation detail.

Reviewer: I would remove the sentence as it provides no further relevant information.

> A number of methods in `SessionHandler` are now deprecated and a result in a noop.

Reviewer 2: ```suggestion
A number of methods in `SessionHandler` are now deprecated and result in a noop.
```

> This mostly includes methods that have been used to bootstrap the session, such as `setHasValidCookie()`.

Reviewer: ```suggestion
This change mostly affects methods that have been used to bootstrap the session, such as `setHasValidCookie()`.
```

> Additionally accessing the following keys on the session is deprecated.

Reviewer: ```suggestion
Additionally, accessing the following keys on the session is deprecated.
```

They directly map to an existing method in another class and any uses can easily be updated:
- `ipAddress`
- `userAgent`
- `requestURI`
- `requestMethod`
- `lastActivityTime`

Refer to [the implementation](https://github.com/WoltLab/WCF/blob/439de4963c947c3569a0c584f795245f693155b0/wcfsetup/install/files/lib/system/session/SessionHandler.class.php#L168-L178) for details.

### Cookies

> The `_userID`, `_password` and `_cookieHash` cookies will no longer be created and no longer be consumed.

Reviewer: ```suggestion
The `_userID`, `_password` and `_cookieHash` cookies will no longer be created nor consumed.
```

### Virtual Sessions

The virtual session logic existed to support multiple devices per single session in `wcf1_session`.
Virtual sessions are no longer required with the refactored session handling.

Anything related to virtual sessions has been completely removed as they are considered an implementation detail.
This removal includes PHP classes and database tables.

### Security Token Constants

The security token constants are deprecated.

> Instead the methods on `SessionHandler` should be used (e.g. `->getSecurityToken()`).

Reviewer: ```suggestion
Instead, the methods of `SessionHandler` should be used (e.g. `->getSecurityToken()`).
```

> Within templates it is recommended to use the `{csrfToken}` tag in place of `{@SECURITY_TOKEN_INPUT_TAG}`.

Reviewer: ```suggestion
Within templates, it is recommended to use the `{csrfToken}` tag in place of `{@SECURITY_TOKEN_INPUT_TAG}`.
```

Is it only recommended or should it be done. I would word it more strongly in the second case.

> The `{csrfToken}` tag is a drop-in replacements and was backported to WoltLab Suite 5.2+, allowing you to maintain compatibility across a broad range of versions.

Reviewer: ```suggestion
The `{csrfToken}` tag is a drop-in replacement and was backported to WoltLab Suite 5.2+, allowing you to maintain compatibility across a broad range of versions.
```
